python/pyarrow/src/arrow/python/udf.cc

```diff
@@ -105,21 +167,109 @@ Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback
   }
   compute::OutputType output_type(options.output_type);
   auto udf_data = std::make_shared<PythonUdf>(
-      wrapper, std::make_shared<OwnedRefNoGIL>(user_function), options.output_type);
+      std::make_shared<OwnedRefNoGIL>(user_function), wrapper, options.output_type);
   compute::ScalarKernel kernel(
       compute::KernelSignature::Make(std::move(input_types), std::move(output_type),
                                      options.arity.is_varargs),
-      PythonUdfExec);
+      PythonUdfExec, kernel_init);
   kernel.data = std::move(udf_data);
 
   kernel.mem_allocation = compute::MemAllocation::NO_PREALLOCATE;
   kernel.null_handling = compute::NullHandling::COMPUTED_NO_PREALLOCATE;
   RETURN_NOT_OK(scalar_func->AddKernel(std::move(kernel)));
-  auto registry = compute::GetFunctionRegistry();
+  if (registry == NULLPTR) {
+    registry = compute::GetFunctionRegistry();
+  }
   RETURN_NOT_OK(registry->AddFunction(std::move(scalar_func)));
   return Status::OK();
 }
 
+}  // namespace
+
+Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback wrapper,
+                              const ScalarUdfOptions& options,
+                              compute::FunctionRegistry* registry) {
+  return RegisterScalarLikeFunction(
+      user_function,
+      PythonScalarUdfKernelInit{std::make_shared<OwnedRefNoGIL>(user_function)}, wrapper,
+      options, registry);
+}
+
+Status RegisterTabularFunction(PyObject* user_function, ScalarUdfWrapperCallback wrapper,
+                               const ScalarUdfOptions& options,
+                               compute::FunctionRegistry* registry) {
+  if (options.arity.num_args != 0 || options.arity.is_varargs) {
+    return Status::Invalid("tabular function must have no arguments");
+  }
+  return RegisterScalarLikeFunction(
```

Comment:
Shouldn't we validate the the `out_type` is a StructType?

Reply:
I'm not sure what you mean; I don't see `out_type` here.

Reply:
OK, I see what you mean now. I'll add a check.